lib/src/multires_stochastic_texture_synthesis.rs

```diff
@@ -1018,7 +987,49 @@ impl Generator {
     }
 }
 
-fn k_neighs_to_color_pattern(
+#[inline]
+fn metric_cauchy(a: u8, b: u8, sig2: f32) -> f32 {
+    let mut x2 = (f32::from(a) - f32::from(b)) / 255.0; //normalize the colors to be between 0-1
+    x2 = x2 * x2;
+    (1.0 + x2 / sig2).ln()
+}
+
+#[inline]
+fn metric_l2(a: u8, b: u8) -> f32 {
+    let x = (f32::from(a) - f32::from(b)) / 255.0;
+    x * x
+}
+
+#[inline]
+fn get_color_of_neighbor(
+    outside_color: image::Rgba<u8>,
+    source_maps: &[ImageBuffer<'_>],
+    n_coord: SignedCoord2D,
+    n_map: MapId,
+    neighbor_color: &mut [u8],
+    is_wrap_mode: bool,
+    wrap_dim: (i32, i32),
+) {
+    let coord = if is_wrap_mode {
+        n_coord.wrap(wrap_dim)
+    } else {
+        n_coord
+    };
+
+    //check if he haven't gone outside the possible bounds
+    if source_maps[n_map.0 as usize].is_in_bounds(coord) {
+        neighbor_color.copy_from_slice(
+            &(source_maps[n_map.0 as usize])
+                .get_pixel(coord.x as u32, coord.y as u32)
+                .0[..4],
+        );
+    } else {
+        // if we have gone out of bounds, then just fill as outside color
+        neighbor_color.copy_from_slice(&outside_color.0[..]);
+    }
+}
+
+fn k_neighs_to_precomputed_reference_pattern(
```

Comment:
renamed this to differentiate it from the now on the fly read candidate pixels, this function is now only called to build `my_pattern` and `my_guide_pattern`